Refresh zone after record is deleted

sauerj opened the pull request:
After the ACME challenge is removed from the zone, the new zone data
must be applied to the DNS servers.

sauerj committed:
Refresh zone after record is deleted

After the ACME challenge is removed from the zone, the new zone data
must be applied to the DNS servers.

sauerj committed:
Fix refresh action

A syntax and variable name error messed up the refresh action.

sauerj commented:
Sorry, I messed up this pull request. Fixes are on the way.

sauerj commented:
Now everything should work.

mcdado commented on bin/delete-record.js:
```diff
@@ -37,6 +37,13 @@ ovh.request('GET', `/domain/zone/${dom.domain}.${dom.tld}/record`, {
+      }
+    });
+
     process.exit(0);
```
Thank you again for contributing!
Unfortunately it doesn't work yet, the async POST call to refresh the zone never happens because `process.exit(0)` is called before. You could simply delete that line, since it's completely redundant. Also, I personally like to use different exit codes, so I know where the process exited. You could simply go up a number and use `process.exit(3)` instead of `process.exit(1)`.

sauerj replied:
Hey, I am completely new to node.js...
I see, the POST call is actually inside the DELETE call. That was *not* what I intended to do.

About the exit code, in 'bin/create_record.js' you used `process.exit(1)` for exiting when a error occurred. I overlooked that you use different exit codes in 'delete_record.js'.

mcdado replied:
We've all been there! ;)

Well, it is actually good the way you did it. You have to wait for the call back from the DELETE call.
The problem is that `ovh.request('POST',...` puts an async operation on the event loop to not block the thread, meaning it continues the flow, hitting `process.exit(0)` and closing the whole program all together

mcdado replied:
I think simply removing `process.exit(0);` will be enough.

sauerj committed:
Call zone refresh in own request

To ensure that the zone refresh call is actually executed, call it from
an own request.

sauerj committed:
Change exit code when zone refresh fails

To distinguish exit point, change exit code when the zone refresh
command fails.

mcdado commented on bin/delete-record.js:
```diff
@@ -36,7 +36,14 @@ ovh.request('GET', `/domain/zone/${dom.domain}.${dom.tld}/record`, {
+    }
   });
+
+  process.exit(0);
```
Well, now this is even worse 😅
Basically what will happen is:

1. DELETE request is scheduled
2. POST request is scheduled
3. Process exited (0)
4. DELETE & POST request are executed at the same time.

Of course 4 never happens because the process is already exited. The way it was done before it was perfect. Before, the flow was:

1. DELETE is scheduled
2. DELETE is executed
3. Callback of DELETE is executed
4. POST is scheduled
5. Process is exited (0)
6. POST is executed.

The problem here was simply that the process exited before the refresh call could be executed. I suggest to go back to the previous commits, simply deleting the `process.exit(0);` line that is superfluous and actually bad.

sauerj replied:
Okay, to be clear: you wrote the `process.exit(0);` not me. That actually confused me a bit.
What confuses me a little bit more, the code actually works fine. :D

mcdado replied:
I know, I wrote it, but when I did there was nothing else to do… while now you have also refresh the zone — which btw is a good idea because it is a necessary step to update the zone.

In my tests it actually doesn't delete the record… you can test by running dig against the DNS record or using the API console here: https://api.ovh.com/console/

sauerj replied:
Okay, I have to excuse. You are right, the record will not be deleted. It is possible that I ran the test with older code... oops.

sauerj committed:
Revert "Call zone refresh in own request"

This reverts commit fa10803900d7c7596a36c1b5ea3f1aa9930cc054.

Commit fa10803 made the situation much worth. We have to wait before
the delete request ist executed before the refresh request can be
sscheduled.

sauerj committed:
Do not exit 'delete-record.js' prematurely

We have to wait until the delete request is finished before exiting the
process.

mcdado commented:
Thank you again for your contributions!
